# Exemplos de uso

### Exemplo de transferência de aprendizado utilizando PyTorch

Vamos seguir o exemplo apresentado por Vasilev et al. (2019), que aplicou uma rede pré-treinada com ImageNet em imagens CIFAR-10.

Inicialmente, vamos importar as bibliotecas:

***Importando biblioteca Torch e TorchVision***

In [1]:
import torch
import torch.nn as nn
import torch.optim as optim
import torchvision
from torchvision import models, transforms

Em seguida, vamos definir o tamanho do lote (batch_size):

***Definindo tamanho do lote***

In [2]:
batch_size = 50

Para definir o conjunto de dados de treinamento, vamos levar em consideração que as imagens CIFAR-10 têm tamanho 32x32, enquanto as redes ImageNet necessitam de dados de tamanho 224x224. 

Então, vamos aumentar as imagens do CIFAR-10 de 32x32 para 224x224. 

É necessário padronizar os dados do CIFAR-10 utilizando a média e o desvio-padrão da ImageNet, e é necessário adicionar um pequeno aumento de dados (flip):

***Definindo dados de treinamento***

In [3]:
# Dados de Treinamento
train_data_transform = transforms.Compose([
    transforms.Resize(224),
    transforms.RandomHorizontalFlip(),
    transforms.RandomVerticalFlip(),
    transforms.ToTensor(),
    transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225])
])

train_set = torchvision.datasets.CIFAR10(root='./data', train=True, download=True, transform=train_data_transform)

train_loader = torch.utils.data.DataLoader(train_set, batch_size=batch_size, shuffle=True, num_workers=2)

Files already downloaded and verified


Realizar os seguintes passos com os dados de validação e teste:

### Definindo dados de teste

In [4]:
val_data_transform = transforms.Compose([
    transforms.Resize(224),
    transforms.ToTensor(),
    transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225])
])

val_set = torchvision.datasets.CIFAR10(root='./data', train=False, download=True, transform=val_data_transform)

val_order = torch.utils.data.DataLoader(val_set, batch_size=batch_size, shuffle=False, num_workers=2)

Files already downloaded and verified


Escolher um dispositivo, preferencialmente GPU:

### Definindo GPU

In [5]:
device = torch.device('cuda:0' if torch.cuda.is_available() else 'cpu')

Definir o modelo de treinamento. Ao contrário do Keras, com PyTorch é necessário realizar as repetições manualmente sobre os dados de treinamento.

Este método repete uma vez por todo o conjunto de treinamento (uma iteração) e aplica um otimizador após cada passo:

### Função de Treinamento do Modelo

In [6]:
def train_model(model, loss_function, optimizer, data_loader):
    
    # Definindo o modelo para o modo de treinamento
    model.train()

    current_loss = 0.0
    current_acc = 0

    # Repete sobre o conjunto de treinamento
    for i, (inputs, labels) in enumerate(data_loader):
        # Envia a entrada ou os rótulos para a GPU
        inputs = inputs.to(device)
        labels = labels.to(device)

        # reinicia os parâmetros do gradiente
        optimizer.zero_grad()

        with torch.set_grad_enabled(True):
            # Passo adiante
            outputs = model(inputs)
            _, predictions = torch.max(outputs, 1)
            loss = loss_function(outputs, labels)

            # passo para trás
            loss.backward()
            optimizer.step()

        # Estatística
        current_loss += loss.item() * inputs.size(0)
        current_acc += torch.sum(predictions == labels.data)

    total_loss = current_loss / len(data_loader.dataset)
    total_acc = current_acc.double() / len(data_loader.dataset)

    print('Train Loss: {:.4f}; Accuracy: {:.4f}'.format(total_loss, total_acc))

### Função de Teste do Modelo

In [7]:
def test_model(model, loss_function, data_loader):

    # Definindo o modelo para o modo de avaliacao
    model.eval()
    current_loss = 0.0
    current_acc = 0
    
     # Repete sobre o conjunto de validação
    for i, (inputs, labels) in enumerate(data_loader):
        # Envia a entrada ou os rótulos para a GPU
        inputs = inputs.to(device)
        labels = labels.to(device)
        
        # Passo adiante
        with torch.set_grad_enabled(False):
            outputs = model(inputs)
            _, predictions = torch.max(outputs, 1)
            loss = loss_function(outputs, labels)
            
        # Estatistica
        current_loss += loss.item() * inputs.size(0)
        current_acc += torch.sum(predictions == labels.data)
        
    total_loss = current_loss / len(data_loader.dataset)
    total_acc = current_acc.double() / len(data_loader.dataset)
    print('Test Loss: {:.4f}; Accuracy: {:.4f}'.format(total_loss, total_acc))

Definir o primeiro cenário de transferência de aprendizado, em que a rede pré-treinada é utilizada como extratora de características. 

Será utilizada uma rede popular, chamada ResNet-18, e o PyTorch fará o download dos pesos prétreinados automaticamente. Vamos substituir a última camada da rede por uma nova camada com 10 classes de saída (uma para cada classe do CIFAR-10). 

Vamos eliminar as camadas da rede original do passo para trás, e atualizar os pesos apenas da camada adicionada utilizando o otimizador ADAM. 

Vamos rodar o treinamento e verificar a acurácia da rede a cada iteração:

### Função de extração de características

In [8]:
def tl_feature_extractor(epochs=3):
    # Carregar o modelo pré-treinado
    model = torchvision.models.resnet18(pretrained=True)

    # Excluir os parâmetros existentes do passo para trás
    for param in model.parameters():
        param.requires_grad = False

    # As novas camadas require, por padrão,
    # requires_grad=True
    num_features = model.fc.in_features
    model.fc = nn.Linear(num_features, 10)

    # Transferindo para a GPU (se disponível)
    
    model = model.to(device)
    
    loss_function = nn.CrossEntropyLoss()
    
    # Otimizar apenas os parâmetros da última camada
    optimizer = optim.Adam(model.fc.parameters())
    
    # treinamento
    for epoch in range(epochs):
        print('Epoch {}/{}'.format(epoch + 1, epochs))
    
    train_model(model, loss_function, optimizer, train_loader)
    test_model(model, loss_function, val_order)

Definir o segundo cenário para refinamento da rede original. Similar ao cenário anterior, porém, toda a rede será treinada novamente:

### Função de Tunning

In [9]:
def tl_fine_tuning(epochs=3):
    # Carregar o modelo pré-treinado
    model = models.resnet18(pretrained=True)

    # Substituir a última camada
    num_features = model.fc.in_features
    model.fc = nn.Linear(num_features, 10)

    # Transferir o modelo para a GPU
    model = model.to(device)

    # Função de Perda
    loss_function = nn.CrossEntropyLoss()

    # Otimização de todos os parâmetros
    optimizer = optim.Adam(model.parameters())

    # treinamento
    for epoch in range(epochs):
        print('Epoch {}/{}'.format(epoch + 1, epochs))

    train_model(model, loss_function, optimizer,train_loader)
    test_model(model, loss_function, val_order)

É possível rodar a rede de acordo com o cenário escolhido. 

I. chamar a função tl_feature_extractor(epochs=5);

II. chamar a função tl_fine_tuning(epochs=5).

### Chamando primeira função

In [10]:
tl_feature_extractor(epochs=5)

Epoch 1/5
Epoch 2/5
Epoch 3/5
Epoch 4/5
Epoch 5/5
Train Loss: 1.0513; Accuracy: 0.6421
Test Loss: 0.7255; Accuracy: 0.7528


### Chamando segunda função

In [11]:
tl_fine_tuning(epochs=5)

Epoch 1/5
Epoch 2/5
Epoch 3/5
Epoch 4/5
Epoch 5/5
Train Loss: 0.8023; Accuracy: 0.7211
Test Loss: 0.7507; Accuracy: 0.7471


### Análise

Para o primeiro cenário, a acurácia obtida foi de 64%, enquanto, no segundo cenário, a acurácia foi de 72%.